### PR TITLE
Finetune model

### PR DESCRIPTION
**Pull request by ivanleomk**
| <a href="https://ellipsis.dev" target="_blank">Ellipsis</a> | :rocket: This PR description was created by [Ellipsis](https://www.ellipsis.dev) for commit 4533fefcfc6d269254368dd4299ef3d6bf039d1f.  | 
|--------|--------|

### Summary:
This PR introduces a new model fine-tuning script and a performance graph generation script, along with model performance data.

**Key points**:
- Added `finetune.py` for fine-tuning a model on the Quora dataset.
- Added `graph.py` for generating performance graphs.
- Included `model_perf.json` and `model_perf_dense.json` containing model performance data.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

File: applications/finetune-quora-embeddings/finetune.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+
+        if curr_eval >= best_eval_perf and iteration - time_step >= 2:
+            print(
+                f"No Improvement detected in training loss. Final loss was {curr_eval}"
+            )
+            break
+        print(
+            f"Iteration {iteration+1}: {curr_eval}, eval: ( {best_eval_perf}, {time_step} )"
+        )
+        if curr_eval > best_eval_perf:
```
Comment (ivanleomk) [inline applications/finetune-quora-embeddings/finetune.py]: Early Stopping condition is here

**Comment by ivanleomk** [comment]
Had to write some custom code for an early stopping condition since the sentence transformers lib doesn't seem to support training till we see no improvement in test loss

File: applications/finetune-quora-embeddings/finetune.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+    print(f"Generated {len(train_examples)} train examples and {len(test_examples)}")
+
+    evaluator = evaluation.BinaryClassificationEvaluator.from_input_examples(
+        test_examples
+    )
+
+    best_eval = (evaluator(model, output_path=None), 0)
+    print(f"Initial Eval: {best_eval}")
+    for iteration in range(6):
```
Comment (jxnl) [inline applications/finetune-quora-embeddings/finetune.py]: parameterize this

File: applications/finetune-quora-embeddings/finetune.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+            )
+            break
+        print(
+            f"Iteration {iteration+1}: {curr_eval}, eval: ( {best_eval_perf}, {time_step} )"
+        )
+        if curr_eval > best_eval_perf:
+            print(f"Previous Eval: {best_eval_perf}, new Eval : {curr_eval}")
+            best_eval = (curr_eval, iteration)
+
+    predictions, test_labels = generate_prediction_labels(
```
Comment (jxnl) [inline applications/finetune-quora-embeddings/finetune.py:237]: why not generate it per epoch? then we have more data

File: applications/finetune-quora-embeddings/finetune.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    )
+
+    best_eval = (evaluator(model, output_path=None), 0)
+    print(f"Initial Eval: {best_eval}")
+    for iteration in range(6):
+        # Train Model
+        model.fit(
+            train_objectives=[(train_dataloader, train_loss)],
+            evaluator=evaluator,
+            epochs=1,
```
Comment (jxnl) [inline applications/finetune-quora-embeddings/finetune.py]: why epochs=1 rather than for n_epochs

File: applications/finetune-quora-embeddings/finetune.py [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+        for metric, function in METRICS.items()
+    }
+
+
+@stub.local_entrypoint()
+def main():
+    import json
+
+    PERCENTAGES = [0.7, 0.9]
+    DATASET_SIZE = 261317
```
Comment (jxnl) [inline applications/finetune-quora-embeddings/finetune.py]: why is this `261317`

File: applications/finetune-quora-embeddings/finetune.py [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+    cache_file = "model_perf.json"
+    try:
+        with open(cache_file, "r") as json_file:
+            model_perf = json.load(json_file)
+    except FileNotFoundError:
+        model_perf = {}
+
+    for item in finetune_model.map(PERCENTAGES):
```
Comment (jxnl) [inline applications/finetune-quora-embeddings/finetune.py]: instead of doing percentage just do 


[1000, 2000, ...]

and jsut double it and plot on log

File: applications/finetune-quora-embeddings/model_perf.json [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        "recall": 0.8626447876447877,
+        "AUC": 0.8763114529033567
+    },
+    "0.5": {
+        "accuracy": 0.891340875994175,
+        "precision": 0.7265252425587897,
+        "recall": 0.8528957528957529,
+        "AUC": 0.8779285474909108
+    },
+    "0.7": {
```
Comment (jxnl) [inline applications/finetune-quora-embeddings/model_perf.json:26]: {
        "accuracy": 0.891340875994175,
        "precision": 0.7265252425587897,
        "recall": 0.8528957528957529,
        "AUC": 0.8779285474909108,
        "model": ...
        "epoch": ...
        "data_size": ...
         ...
    }

Comment (jxnl) [reply applications/finetune-quora-embeddings/model_perf.json:26]: and just make an array

Then you can just 

pd.from_jsonlines()
